proof_of_stake/src/epoched.rs

```diff
@@ -0,0 +1,1270 @@
+    pub fn init_at_offset(
+        value: Data,
+        current_epoch: impl Into<Epoch>,
+        offset: DynEpochOffset,
+        params: &PosParams,
+    ) -> Self {
+        let index = offset.value(params) as usize;
+        Self::init_at_index(value, current_epoch, index)
+    }
+
+    /// Initialize new data at the given index.
+    fn init_at_index(
+        value: Data,
+        current_epoch: impl Into<Epoch>,
+        offset: usize,
+    ) -> Self {
+        let mut data = vec![];
+        for _ in 0..offset {
+            data.push(None);
+        }
+        data.push(Some(value));
+        Self {
+            last_update: current_epoch.into(),
+            data,
+            offset: PhantomData,
+        }
+    }
+
+    /// Find the current value for the given epoch as the sum of delta values at
+    /// and before the current epoch.
+    pub fn get(&self, epoch: impl Into<Epoch>) -> Option<Data> {
+        let epoch = epoch.into();
+        let index: usize = (epoch.sub_or_default(self.last_update)).into();
+        self.get_at_index(index)
+    }
+
+    /// Find the value at the offset from the given epoch as the sum of delta
+    /// values at and before the epoch offset.
+    pub fn get_at_offset(
+        &self,
+        epoch: impl Into<Epoch>,
+        offset: DynEpochOffset,
+        params: &PosParams,
+    ) -> Option<Data> {
+        let offset = offset.value(params);
+        let epoch_at_offset = epoch.into() + offset;
+        let index: usize =
+            (epoch_at_offset.sub_or_default(self.last_update)).into();
+        self.get_at_index(index)
+    }
+
+    /// Iterate the delta values set in the data.
+    pub fn iter(&self) -> impl Iterator<Item = &Data> {
+        self.data.iter().filter_map(|value| value.as_ref())
+    }
+
+    /// Iterate the delta values set in the data together with their epoch.
+    pub fn iter_with_epochs(&self) -> impl Iterator<Item = (&Data, Epoch)> {
+        let last_update = self.last_update;
+        self.data
+            .iter()
+            .enumerate()
+            .filter_map(move |(index, value)| {
+                value.as_ref().map(|value| (value, last_update + index))
+            })
+    }
+
+    /// Find the value at the given index as the sum of delta values at and
+    /// before the index.
+    fn get_at_index(&self, offset: usize) -> Option<Data> {
+        let index = cmp::min(offset, self.data.len());
+        let mut sum: Option<Data> = None;
+        for i in 0..index + 1 {
+            if let Some(next) = self.data.get(i) {
+                // Add current to the sum, if any
+                match (&sum, next) {
+                    (Some(current_sum), Some(next)) => {
+                        sum = Some(current_sum.clone() + next.clone())
+                    }
+                    (None, Some(next)) => sum = Some(next.clone()),
+                    _ => {}
+                };
+            }
+        }
+        sum
```

Comment:
```suggestion
        for next in self.data.iter().take(index + 1) {
            // Add current to the sum, if any
            match (&mut sum, next) {
                (current_sum @ Some(_), Some(next)) => {
                    sum = Some(current_sum.take().unwrap() + next.clone())
                }
                (None, Some(next)) => sum = Some(next.clone()),
                _ => {}
            };
        }
        sum
```

Reply:
There's unused alias on line 398, I removed it here:
```suggestion
        for next in self.data.iter().take(index + 1) {
            // Add current to the sum, if any
            match (&mut sum, next) {
                (Some(_), Some(next)) => {
                    sum = sum.map(|current_sum| current_sum + next.clone())
                }
                (None, Some(next)) => sum = Some(next.clone()),
                _ => {}
            };
        }
        sum
```